kernel/src/fs/sysfs/inode.rs

```diff
@@ -0,0 +1,617 @@
+        Err(Error::new(Errno::EPERM))
+    }
+
+    fn lookup(&self, name: &str) -> Result<Arc<dyn Inode>> {
+        if self.type_() != InodeType::Dir {
+            return Err(Error::new(Errno::ENOTDIR));
+        }
+
+        if name == "." {
+            return Ok(self.this());
+        } else if name == ".." {
+            return Ok(self.parent.upgrade().unwrap_or_else(|| self.this()));
+        }
+        match &self.inner_node {
+            InnerNode::Branch(obj) => {
+                if let Some(branch) = obj.as_branch() {
+                    self.lookup_node_or_attr(name, branch)
+                } else if let Some(node) = obj.as_node() {
+                    self.lookup_attr(name, node)
+                } else {
+                    Err(Error::new(Errno::ENOTDIR))
+                }
+            }
+            _ => Err(Error::new(Errno::ENOTDIR)),
+        }
+    }
+
+    /// Reads directory entries starting from a given offset.
+    ///
+    /// Why interpreting the `offset` argument as an inode number?
+    ///
+    /// It may take multiple `getdents` system calls
+    /// -- and thus multiple calls to this method --
+    /// to list a large directory when the syscall is provided a small buffer.
+    /// Between these calls,
+    /// the directory may have new entries added or existing ones removed
+    /// by some concurrent users that are working on the directory.
+    /// In such situations,
+    /// missing some of the concurrently-added entries is inevitable,
+    /// but reporting the same entry multiple times would be
+    /// very confusing to the user.
+    ///
+    /// To address this issue,
+    /// the `readdir_at` method reports entries starting from a user-given `offset`
+    /// and returns an increment that the next call should be put on the `offset` argument
+    /// to avoid getting duplicated entries.
+    ///
+    /// Different file systems may interpret the meaning of
+    /// the `offset` argument differently:
+    /// one may take it as a _byte_ offset,
+    /// while the other may treat it as an _index_.
+    /// This freedom is guaranteed by Linux as documented in
+    /// [the man page of getdents](https://man7.org/linux/man-pages/man2/getdents.2.html).
+    ///
+    /// Our implementation of sysfs interprets the `offset`
+    /// as an _inode number_.
+    /// By inode numbers, directory entries will have a _stable_ order
+    /// across different calls to `readdir_at`.
+    fn readdir_at(&self, offset: usize, visitor: &mut dyn DirentVisitor) -> Result<usize> {
+        let start_ino = offset as Ino;
+        let mut count = 0;
+        let mut last_ino = start_ino;
+
+        // Collect all entries
+        let mut entries: Vec<_> = self.new_dentry_iter(0).collect();
+
+        // Sort by ino then name for deterministic order
+        entries.sort_by(|a, b| a.ino.cmp(&b.ino).then_with(|| a.name.cmp(&b.name)));
+
+        // Deduplicate by ino, keeping first occurrence
```

Comment:
Sorry, I didn't get why sorting and dedup are needed. The inode numbers are unique. This is guaranteed by the `ino` sub-module and the `SysNodeId` type.